[Merged by Bors] - feat(Topology): implement delaborators for non-standard topology notation

robertmaxton42 opened the pull request:
Add delaborators for unary and binary notation related to non-standard topologies in the TopologicalSpace namespace.


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

robertmaxton42 committed:
Implement delaborators for non-standard topology notation

Added delaborators for unary and binary notation related to non-standard topologies in the TopologicalSpace namespace.

eric-wieser commented on Mathlib/Topology/Defs/Basic.lean:
```diff
@@ -195,6 +195,47 @@ scoped[Topology] notation (name := closure_of) "closure[" t "]" => @closure _ t
+    if (← Meta.isDefEq instα synthInstα) ∧ (← Meta.isDefEq instβ synthInstβ) then failure
+    let instDα ← withNaryArg 2 delab
+    let instDβ ← withNaryArg 3 delab
+    mkStx instDα instDβ
```
I suspect it's not worth extracting any more than is already extracted

robertmaxton42 committed:
Extract common code into assertNonCanonical helper

robertmaxton42 committed:
shorten line

kckennylau commented on Mathlib/Topology/Defs/Basic.lean:
```diff
@@ -195,6 +195,47 @@ scoped[Topology] notation (name := closure_of) "closure[" t "]" => @closure _ t
+open Topology Lean Meta PrettyPrinter.Delaborator SubExpr
+
+/-- Fails iff argument #`space` has a `TopologicalSpace` instance registered and
+that instance is defeq to argument #`inst`. -/
```
```suggestion
/-- Fails iff argument #`space` has a `TopologicalSpace` instance registered and
that instance is defeq to argument #`inst`, which is when we can use the default notation. -/
```

kckennylau commented on Mathlib/Topology/Defs/Basic.lean:
```diff
@@ -195,6 +195,47 @@ scoped[Topology] notation (name := closure_of) "closure[" t "]" => @closure _ t
+  let α ← withNaryArg space getExpr
+  let .some synthInst ← Meta.trySynthInstance (← Meta.mkAppM ``TopologicalSpace #[α]) | return ()
+  let inst ← withNaryArg inst getExpr
+  if ← Meta.isDefEq inst synthInst then failure
```
```lean
import Mathlib.Topology.Defs.Basic

namespace TopologicalSpace
open Topology Lean Meta PrettyPrinter.Delaborator SubExpr

/-- Given an expression for an instance, check if it is equal to the automatically synthesized one
(if it exists), and returns the delaborated instance **iff** the synthesized one is different
or does not exist. -/
def delabNonCanonical : Delab := do
  let instD ← delab
  let inst ← getExpr
  let type ← inferType inst
  -- if there is no synthesized instance, return the one in the expression
  -- (because it is still non-canonical)
  let .some synthInst ← Meta.trySynthInstance type | return instD
  -- if they are defeq, then use the default delab
  let .false ← Meta.isDefEq inst synthInst | failure
  return instD

/-- Delaborate unary notation referring to non-standard topologies. -/
def delabUnary (mkStx : Term → DelabM Term) : Delab :=
  withOverApp 2 <| whenPPOption Lean.getPPNotation do
    let instD ← withNaryArg 1 delabNonCanonical
    mkStx instD

/-- Delaborate binary notation referring to non-standard topologies. -/
def delabBinary (mkStx : Term → Term → DelabM Term) : Delab :=
  withOverApp 4 <| whenPPOption Lean.getPPNotation do
    -- fall through to normal delab if both canonical
    let instDα ← withNaryArg 2 delabNonCanonical
    let instDβ ← withNaryArg 3 delabNonCanonical
    mkStx instDα instDβ

/-- Delaborator for `IsOpen[_]`. -/
@[app_delab IsOpen] def delabIsOpen : Delab := delabUnary fun x ↦ `(IsOpen[$x])

/-- Delaborator for `IsClosed[_]`. -/
@[app_delab IsClosed] def delabIsClosed : Delab := delabUnary fun x ↦ `(IsClosed[$x])

/-- Delaborator for `closure[_]`. -/
@[app_delab closure] def delabClosure : Delab := delabUnary fun x ↦ `(closure[$x])

/-- Delaborator for `Continuous[_, _]`. -/
@[app_delab Continuous] def delabContinuous : Delab := delabBinary fun x y ↦ `(Continuous[$x, $y])


section RemoveMe

variable {α : Type*} [σ : TopologicalSpace α]
  (f : TopologicalSpace α → TopologicalSpace α) (s : Set α)

example : IsOpen[σ] s := by
  sorry 

example : IsOpen[f σ] s := by
  sorry 

example : IsOpen[σ] = sorry := by
  sorry 

example : IsOpen[f σ] = sorry := by
  sorry 

end RemoveMe

end TopologicalSpace
```

kckennylau commented on Mathlib/Topology/Defs/Basic.lean:
```diff
@@ -195,6 +195,47 @@ scoped[Topology] notation (name := closure_of) "closure[" t "]" => @closure _ t
+  withOverApp 2 <| whenPPOption Lean.getPPNotation do
+    assertNonCanonical 0 1 -- fall through to normal delab if canonical
+    let instD ← withNaryArg 1 delab
+    mkStx instD
```
For the purpose of further extensibility (i.e. code reusing in other parts of the library), it might be worthwhile to consider not hardcoding the inputs `2` and `1` (in my new implementation there is no `0`).

robertmaxton42 commented on Mathlib/Topology/Defs/Basic.lean:
```diff
@@ -195,6 +195,47 @@ scoped[Topology] notation (name := closure_of) "closure[" t "]" => @closure _ t
+  withOverApp 4 <| whenPPOption Lean.getPPNotation do
+    -- fall through to normal delab if both canonical
+    assertNonCanonical 0 2 <|> assertNonCanonical 1 3
+    let instDα ← withNaryArg 2 delab
```
Unfortunately, doing that means that we commit to failing if _this particular_ instance is non-canonical -- I don't think we can use `<|>` anymore, so behavior changes from NAND (not both canonical, at least one is non-canonical) to NOR (neither is canonical, succeeds only if both are non-canonical).

Everything else looks reasonable, though, so I'll try and bundle up a next iteration with it, thanks!

robertmaxton42 replied:
(Also, the resulting `assertNonCanonical` is no longer bound to `TopologicalSpace` and belongs somewhere more general; suggestions?)

robertmaxton42 committed:
Refactor core delab helper as monadic read-and-check function

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

robertmaxton42 committed:
Create tests for delabs

Add tests for induced and coinduced properties in topology.

robertmaxton42 committed:
Marginally generalize checks

robertmaxton42 committed:
bump proofwidgets

Updated ProofWidgets version to v0.0.80.

robertmaxton42 committed:
update toolchain

robertmaxton42 committed:
Revert "update toolchain"

This reverts commit 5a6e10997634d3d327900c9ebe5d2898ca0aac57.

robertmaxton42 committed:
update proofwidgets

robertmaxton42 committed:
Merge branch 'master' into delab_isFoo_of

kckennylau commented on MathlibTest/Delab/TopologicalSpace.lean:
```diff
@@ -0,0 +1,33 @@
+import Mathlib.Topology.Order
```
I see, it's because of the bug right (might be worth explaining this in a comment too)

robertmaxton42 replied:
`synthInstance` and friends don't work properly in `#check`, `#guard_target` when used with opaque typeclass assumptions/variable definitions; we need an actual concrete type with an actual concrete canonical instance in order to make a linter that checks what we want, and it's actually kinda hard to make a `TopologicalSpace` instance without importing a whole bunch of the `Set` API anyway. At which point, we might as well just import `Mathlib.Topology.Order`, which gives us a topology on `Nat` we can use.

kckennylau commented on MathlibTest/Delab/TopologicalSpace.lean:
```diff
@@ -0,0 +1,33 @@
+
+/-- info: Continuous[induced p inferInstance, coinduced p inferInstance] p ↔ Continuous p : Prop -/
+#guard_msgs(info) in
+#check Continuous[induced p inferInstance, coinduced p inferInstance] p ↔ Continuous[inferInstance, inferInstance] p
```
Maybe also add tests for the `IsOpen[..]` thing without the set? Since you specifically wanted that to work too. (Though as I've explained, I don't know why you would want that)

kckennylau commented on Mathlib/Topology/Defs/Basic.lean:
```diff
@@ -199,6 +199,50 @@ scoped[Topology] notation (name := closure_of) "closure[" t "]" => @closure _ t
+  -- if there is no synthesized instance, still return `false`
+  -- (because `inst` is still non-canonical)
+  let .some synthInst ← Meta.trySynthInstance type | return (false, instD)
+  return (← Meta.isDefEq inst synthInst, instD)
```
maybe this can be moved to Util?

kckennylau commented on Mathlib/Topology/Defs/Basic.lean:
```diff
@@ -199,6 +199,50 @@ scoped[Topology] notation (name := closure_of) "closure[" t "]" => @closure _ t
+    -- fall through to normal delab if both canonical
+    let (canonα?, instDα) ← withNaryArg 2 delabCheckingCanonical
+    let (canonβ?, instDβ) ← withNaryArg 3 delabCheckingCanonical
+    if canonα? && canonβ? then failure
```
```suggestion
    let (canonα?, instDα) ← withNaryArg 2 delabCheckingCanonical
    let (canonβ?, instDβ) ← withNaryArg 3 delabCheckingCanonical
    -- fall through to normal delab if both canonical
    if canonα? && canonβ? then failure
```

kckennylau commented on Mathlib/Topology/Defs/Basic.lean:
```diff
@@ -199,6 +199,50 @@ scoped[Topology] notation (name := closure_of) "closure[" t "]" => @closure _ t
+    -- fall through to normal delab if both canonical
+    let (canonα?, instDα) ← withNaryArg 2 delabCheckingCanonical
+    let (canonβ?, instDβ) ← withNaryArg 3 delabCheckingCanonical
+    if canonα? && canonβ? then failure
```
I have a cursed idea. What if if `canonα? = true` and `canonβ? = false` then we print ``mkStx `(_) instDβ``

kckennylau replied:
The new hole doesn't have the correct syntax information, and I don't know how to fix this. Ideally we want the hole to be with the syntax of the instance when you hover over it.

kckennylau replied:
I've cracked the case:
```suggestion
    mkStx (if canonα? then .mk (mkHole instDα) else instDα)
      (if canonβ? then .mk (mkHole instDβ) else instDβ)
```

robertmaxton42 replied:
Can you elaborate on why you want this?

eric-wieser replied:
No, you should be emiting the hole inside with `withNaryArg` so that the correct expr is attached.

robertmaxton42 committed:
Add DelabM helper for instance synthesis checking

robertmaxton42 committed:
Move all delab helpers to Util

robertmaxton42 committed:
add opens

robertmaxton42 committed:
use imported util helpers

robertmaxton42 committed:
mk_all

robertmaxton42 committed:
avoid ℕ notation

Updated type annotations from `ℕ` to `Nat` for delabUnary and delabBinary functions.

robertmaxton42 committed:
finish delabBinary

robertmaxton42 committed:
Update Basic.lean

robertmaxton42 committed:
long line linter

eric-wieser commented on MathlibTest/Delab/TopologicalSpace.lean:
```diff
@@ -0,0 +1,33 @@
+
+/-- info: IsOpen[induced f inferInstance] s ↔ IsOpen s : Prop -/
+#guard_msgs(info) in
+#check IsOpen[induced f inferInstance] s ↔ IsOpen[inferInstance] s
```
I think we do want exactly Kenny's test-case, with a TODO.

Optionally we could have a very simple check for "if the instance is a free variable and the local context contains another free variable of the same time, mark it as non-canonical.

eric-wieser commented on MathlibTest/Delab/TopologicalSpace.lean:
```diff
@@ -0,0 +1,33 @@
+
+variable {α : Type*} [TopologicalSpace α] (f : ℕ → α) (p : ℕ → Prop) (s : Set ℕ)
+
+-- For unary ops, we check that `induced p inferInstance` triggers the delab and `inferInstance` does not
```
If these comments describe multiple tests, consider using `section` to make the grouping explicit.

eric-wieser commented on Mathlib/Util/DelabNonCanonical.lean:
```diff
@@ -0,0 +1,50 @@
+  let type ← inferType inst
+  -- if there is no synthesized instance, still return `false`
+  -- (because `inst` is still non-canonical)
+  let .some synthInst ← Meta.trySynthInstance type | return (false, instD)
```
This function seems to be doing unrelated two unrelated things; why not
```lean
def Lean.Meta.isCanonicalInstance (inst : Expr) : MetaM Bool := do
  let type ← inferType inst
  let .some synthInst ← Meta.trySynthInstance type |
    -- no global instance means it can't be canonical
    return false
  return ← Meta.isDefEq inst synthInst
```
then
```lean
    mkStx <| ← withNaryArg arg do
      guard !(← Meta.isCanonicalInstance (← getExpr))
      delab
```
when you need it

eric-wieser replied:
You probably also should be using instance reducibility

eric-wieser commented on Mathlib/Util/DelabNonCanonical.lean:
```diff
@@ -0,0 +1,50 @@
+`arg` is a non-canonical instance (is not defeq to what is synthesized for its type, or else
+instance synthesis fails). -/
+def delabUnary (arity arg : Nat) (mkStx : Term → DelabM Term) : Delab :=
+  withOverApp arity <| whenPPOption Lean.getPPNotation do
```
I think these belong at the caller, not here.

Vierkantor replied:
As far as I can tell, `pp.explicit` doesn't seem to appear in the definition of [`getPPNotation`](https://github.com/leanprover/lean4/blob/714601baf118066cbf3f282361339c6d06665b2a/src/Lean/PrettyPrinter/Delaborator/Options.lean#L254)?

```lean
def getPPAll (o : Options) : Bool := o.get pp.all.name false
...
def getPPNotation (o : Options) : Bool := o.get pp.notation.name (!getPPAll o)
```

Vierkantor replied:
The problem with putting the `withOverApp` and `whenPPOption` here is the caller loses quite a bit of flexibility. It seems we don't always call `withOverApp` or `whenPPOption` in consistent ways throughout Mathlib (e.g. `iUnion_delab` doesn't do `withOverApp` but `iSup_delab` does). @eric-wieser: do you have an example where this would be intentional rather than a mistake?

eric-wieser replied:
I think the difference is that `iSup` can be applied to pi types, but `iUnion` cannot; so strictly speaking, `iUnion_delab` sounds broken, but only in cases that abuse the defeq of set membership.

eric-wieser replied:
My argument for putting `withOverApp` at the caller is that the reader can immediately say "yes, this follows the standard pattern" without having to look at further definitions.

eric-wieser replied:
At any rate, I do think we should be skipping the notation if `pp.explicit` is set to `true`; we don't need `IsOpen[t]` if the user has asked for `@IsOpen X t`.

robertmaxton42 replied:
@Vierkantor Sorry, I'm not seeing how `withOverApp` loses flexibility here? I agree with @eric-wieser that `pp.explicit` should definitely be 'hard-coded'

Also bah, I misread the definition somehow. Let me fix that.

robertmaxton42 replied:
```suggestion
  whenPPOption Lean.getPPNotation <| whenNotPPOption getPPExplicit <| withOverApp arity <| do
```

eric-wieser replied:
Unresolving for @Vierkantor to comment further on "loses flexibility"

eric-wieser commented on MathlibTest/Delab/TopologicalSpace.lean:
```diff
@@ -0,0 +1,33 @@
+import Mathlib.Topology.Order
+
+open TopologicalSpace
+open scoped Topology
```
You should also test that the delaborator isn't used when the scope is not open.

robertmaxton42 committed:
check no instance, unapplied

robertmaxton42 committed:
fix scoping, check scoping

robertmaxton42 committed:
multiple instances check

robertmaxton42 committed:
Merge branch 'master' into delab_isFoo_of

robertmaxton42 committed:
use instance reducibility, probably

j-loreaux reviewed (commented):
@eric-wieser is there anything else you want to happen here? I would like to get this across the line soon.

robertmaxton42 committed:
Merge branch 'leanprover-community:master' into delab_isFoo_of

robertmaxton42 committed:
make canonical instances appear as `_`

robertmaxton42 committed:
linters, cleanup

eric-wieser commented on Mathlib/Util/DelabNonCanonical.lean:
```diff
@@ -0,0 +1,58 @@
+def delabCheckingCanonical : DelabM (Bool × Term) := do
+  let inst ← getExpr
+  if ← isCanonicalInstance inst then
+    return (true, ← annotateTermInfo <| ← `(_))
```
Should this be `withAnnotateTermInfo` as I suggested above?

robertmaxton42 committed:
cleanup, withAnnotateTermInfo

j-loreaux commented:
maintainer merge

robertmaxton42 committed:
Update Mathlib/Util/DelabNonCanonical.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

robertmaxton42 committed:
Merge branch 'master' into delab_isFoo_of

Vierkantor commented:
bors d=@eric-wieser

eric-wieser commented on MathlibTest/Delab/TopologicalSpace.lean:
```diff
@@ -0,0 +1,178 @@
+
+-- Note: This is *not* the desired behavior. Treat it instead as a 'bug tracker.'
+-- When `(try)SynthInstance` in `$check` is fixed, this should no longer trigger the delab
+-- on the right, so we should get `Continuous[induced g inferInstance, _]`.
```
This is a bit wordy; let's start with ``-- TODO: should be `Continuous[induced g inferInstance, _]`` on the first line, and explain why on the later lines.

robertmaxton42 replied:
```suggestion
-- TODO: should be `Continuous[induced g inferInstance, _]
-- When `(try)SynthInstance` in `#check` is fixed, the delab should no longer trigger on the right.
```

eric-wieser commented on MathlibTest/Delab/TopologicalSpace.lean:
```diff
@@ -0,0 +1,178 @@
+-- in either or both slots do trigger the delab and that
+-- `instTopologicalSpaceNat`, `sierpinskiSpace` does not.
+
+/-- info: [Continuous[τ₁, _] p, Continuous[τ₂, _] p, Continuous[induced g inferInstance, inferInstance] g] : List Prop -/
```
Don't we expect
```suggestion
/-- info: [Continuous[τ₁, _] p, Continuous[τ₂, _] p, Continuous[induced g inferInstance, _] g] : List Prop -/
```
here?

robertmaxton42 replied:
Sort of. Same problem as the other TODOs. Corrected, thanks!

robertmaxton42 commented on Mathlib/Util/DelabNonCanonical.lean:
```diff
@@ -0,0 +1,58 @@
+argument `arg₁` or `arg₂` are non-canonical instances (are not defeq to what is synthesized for
+its type, or else instance synthesis fails). -/
+def delabBinary (arity arg₁ arg₂ : Nat) (mkStx : Term → Term → DelabM Term) : Delab :=
+  withOverApp arity <| whenPPOption Lean.getPPNotation do
```
```suggestion
  whenPPOption Lean.getPPNotation <| whenNotPPOption getPPExplicit <| withOverApp arity <| do
```

robertmaxton42 commented on MathlibTest/Delab/TopologicalSpace.lean:
```diff
@@ -0,0 +1,178 @@
+
+-- As above, this is not the desired behavior. When `(try)SynthInstance` in `#check` is fixed,
+-- this should no longer trigger the delab on the left, so we should get
+-- `Continuous[_, coinduced h inferInstance]`.
```
```suggestion
-- TODO should be `Continuous[_, coinduced h inferInstance]`.
-- When `(try)SynthInstance` in `#check` is fixed, the delab should no longer trigger on the left.
```

robertmaxton42 commented on MathlibTest/Delab/TopologicalSpace.lean:
```diff
@@ -0,0 +1,178 @@
+/-- info: [Continuous[τ₁, _] p, Continuous[τ₂, _] p, Continuous[induced g inferInstance, inferInstance] g] : List Prop -/
+#guard_msgs(info) in
+#check [Continuous[τ₁, sierpinskiSpace] p, Continuous[τ₂, sierpinskiSpace] p,
+  Continuous[induced g inferInstance, inferInstance] g]
```
```suggestion
/-- info: [Continuous[τ₁, _] p, Continuous[τ₂, _] p] : List Prop -/
#guard_msgs(info) in
#check [Continuous[τ₁, sierpinskiSpace] p, Continuous[τ₂, sierpinskiSpace] p]

-- TODO should be `Continuous[induced g inferInstance, _]`
-- Once `(try)SynthInstance` in `#check` is fixed, the delab should no longer trigger on the right.
/--info: Continuous[induced g inferInstance, inferInstance] g : Prop -/
#guard_msgs(info) in
#check Continuous[induced g inferInstance, inferInstance] g
```

robertmaxton42 commented on MathlibTest/Delab/TopologicalSpace.lean:
```diff
@@ -0,0 +1,178 @@
+/-- info: [Continuous[_, σ₁] p, Continuous[_, σ₂] p, Continuous[inferInstance, coinduced h inferInstance] h] : List Prop -/
+#guard_msgs(info) in
+#check [Continuous[instTopologicalSpaceNat, σ₁] p, Continuous[instTopologicalSpaceNat, σ₂] p,
+  Continuous[inferInstance, coinduced h inferInstance] h]
```
```suggestion
/-- info: [Continuous[_, σ₁] p, Continuous[_, σ₂] p] : List Prop -/
#guard_msgs(info) in
#check [Continuous[instTopologicalSpaceNat, σ₁] p, Continuous[instTopologicalSpaceNat, σ₂] p]

-- TODO should be `Continuous[_, coinduced h inferInstance]`
-- Once `(try)SynthInstance` in `#check` is fixed, the delab should no longer trigger on the left.
/-- info: Continuous[inferInstance, coinduced h inferInstance] h : Prop -/
#guard_msgs(info) in
#check Continuous[inferInstance, coinduced h inferInstance] h
```

robertmaxton42 committed:
Apply suggestions from code review

Co-authored-by: Robert Maxton <robert.maxton42@gmail.com>

eric-wieser commented on Mathlib/Util/DelabNonCanonical.lean:
```diff
@@ -0,0 +1,58 @@
+  let inst ← getExpr
+  if ← isCanonicalInstance inst then
+    return (true, ← withAnnotateTermInfo `(_))
+  else
+    return (false, ← delab)
+namespace Delab.Noncanonical
```
nit:
```suggestion
    return (false, ← delab)

namespace Delab.Noncanonical
```

eric-wieser commented:
I think just [this thread](https://github.com/leanprover-community/mathlib4/pull/31425#discussion_r2688375249) (started Jan 13) to resolve.

eric-wieser committed:
Update Mathlib/Util/DelabNonCanonical.lean

j-loreaux commented:
@Vierkantor I think it's you who needs to respond to the discussion above.

robertmaxton42 committed:
Merge branch 'master' into delab_isFoo_of

robertmaxton42 commented on MathlibTest/Delab/TopologicalSpace.lean:
```diff
@@ -0,0 +1,186 @@
+/-- info:
+Continuous[induced g inferInstance, inst✝] : (β → α) → Prop -/
+#guard_msgs(info) in
+#check Continuous[induced g inferInstance, inferInstanceAs% (TopologicalSpace α)]
```
```suggestion
#check Continuous[induced g inferInstance, (inferInstance : TopologicalSpace α)]
```

robertmaxton42 commented on MathlibTest/Delab/TopologicalSpace.lean:
```diff
@@ -0,0 +1,186 @@
+/-- info:
+Continuous[inst✝, coinduced h inferInstance] : (α → β) → Prop -/
+#guard_msgs(info) in
+#check Continuous[inferInstanceAs% (TopologicalSpace α), coinduced h inferInstance]
```
```suggestion
#check Continuous[(inferInstance : TopologicalSpace α), coinduced h inferInstance]
```

robertmaxton42 committed:
remove inferInstanceAs%

Co-authored-by: Robert Maxton <robert.maxton42@gmail.com>

robertmaxton42 commented on MathlibTest/Delab/TopologicalSpace.lean:
```diff
@@ -0,0 +1,186 @@
+-- TODO: should be `Continuous[induced g inferInstance, _]
+-- When `(try)SynthInstance` in `#check` is fixed, the delab should no longer trigger on the right.
+/-- info:
+Continuous[induced g inferInstance, inst✝] : (β → α) → Prop -/
```
```suggestion
Continuous[induced g inferInstance, inferInstance] : (β → α) → Prop -/
```

robertmaxton42 commented on MathlibTest/Delab/TopologicalSpace.lean:
```diff
@@ -0,0 +1,186 @@
+-- TODO should be `Continuous[_, coinduced h inferInstance]`.
+-- When `(try)SynthInstance` in `#check` is fixed, the delab should no longer trigger on the left.
+/-- info:
+Continuous[inst✝, coinduced h inferInstance] : (α → β) → Prop -/
```
```suggestion
Continuous[inferInstance, coinduced h inferInstance] : (α → β) → Prop -/
```

robertmaxton42 committed:
`inst✝`s are `inferInstance`s again

Co-authored-by: Robert Maxton <robert.maxton42@gmail.com>
